File: haskell/rna-transcription/DNA.hs
module DNA (toRNA) where

import qualified Data.Map as Map

type Nucleotide = Char
type Strand = [Nucleotide]

dnaToRna :: Nucleotide -> Nucleotide
dnaToRna nucleotide =
  case Map.lookup nucleotide dnaRnaMap of
    Just rna -> rna
    Nothing -> error "Not a valid DNA nucleotide"
  where dnaRnaMap = Map.fromList $ zip "GCTA" "CGAU"

toRNA :: Strand -> Strand
toRNA = map dnaToRna

File: python/rna-transcription/dna.py
_dna_to_rna_map = {
    'G': 'C',
    'C': 'G',
    'T': 'A',
    'A': 'U',
}


def to_rna(strand):
    """ Given a DNA strand, returns its RNA complement (per RNA transcription).
    """
    return ''.join(_dna_to_rna_map[nucleotide]
                   for nucleotide in strand)
